### ISSUE TITLE
`wp (api|json) generate application-password*`

### ISSUE THREAD

**Issue by pfefferle** [wp-cli/ideas#160]
## Feature Request

- [X] Yes, I reviewed the [contribution guidelines](https://make.wordpress.org/cli/handbook/contributing/).

**Describe your use case and the problem you are facing**

A new cli command to handle API commands in general and a command to generate `application-passwords` in particular.

**Describe the solution you'd like**

To setup a WordPress blog using the CLI, it would be helpful to also have a command to pre-generate `application-passwords`.

**Comment by swissspidy** [comment]
Since app passwords are stored per-user, this would fall under the `wp user` umbrella ([wp-cli/entity-command](https://github.com/wp-cli/entity-command)).

Since app passwords can be [added, listed and revoked](https://make.wordpress.org/core/2020/11/05/application-passwords-integration-guide/), this could result in a new family of `wp user` sub commands.

In the meantime, you might be able to use `wp eval` for this (untested):

```bash
$  wp eval 'WP_Application_Passwords::create_new_application_password( 123, array( "name" => "test" ) )'
```

**Comment by pfefferle** [comment]
Good point, thanks for your answer @swissspidy !

Should I move the issue to the `entity-command` repo?

**Comment by swissspidy** [comment]
We'll transfer it if needed. Can keep it here for discussion :)

**Comment by DeLm0re** [comment]
Are there any news about this suggestion ?

After reading some documentation, and like @swissspidy said, the only working solution for now is to use the following command :

`$  wp eval 'WP_Application_Passwords::create_new_application_password( 123, array( "name" => "test" ) );'`

Note that the `;` is **mandatory** otherwise it won't work in command line.

There is still a main issue for me with this solution.
In order to deploy WordPress websites accessible from my own tier application. I have to store generated password. However, **the command above only create an application password and do not give the value !**. And it does not seems to accept a given value for the password. eg: 

`$  wp eval 'WP_Application_Passwords::create_new_application_password( 123, array( "name" => "test", "value" => "8lJB MyXY s3FC ArTF 81Kb QVAe") );'`

In conclusion, it would be very helpful if such a WP CLI command exists: 

`wp generate application-password <userId> <name> <password>`

**Comment by swissspidy** [comment]
> I do already tried something like :
> 
> `$ wp eval '$tmp = WP_Application_Passwords::create_new_application_password( 1, array( "name" => "test" ) ); echo($tmp[1]["password"]);'`
> 
> Unfortunatly, this method returns the **hashed** password ... not the real password value.

Check the method's documentation; the unhashed password is the first array item.

So:

```bash
$ wp eval '$tmp = WP_Application_Passwords::create_new_application_password( 1, array( "name" => "test" ) ); echo($tmp[0]);'
```

-----

As for implementing this feature, I believe we are looking for the following commands:

* `wp user application-password list <user>`
* `wp user application-password get <user> <uuid>`
* `wp user application-password has <user> <name>`
* `wp user application-password update <user> <uuid> [--<field>=<value>]`
* `wp user application-password record-usage <user> <uuid>`
* `wp user application-password create <user> <name> [--app_id=<value>]`

**Comment by DeLm0re** [comment]
@swissspidy Oh wow, thanks for the fast reply.

That is exactly what i'm looking for. Perfect 👍

**Comment by schlessera** [comment]
@swissspidy I'm currently working on implementing these new commands and going by what you've suggested above.

I'm surprised about the need to provide a user ID when retrieving an application password by UUID, which seems to also be a requirement by Core's API. Given a UUID should be unique even across users, I don't really understand the design of the data model... Was that for performance reasons that they are stored in user meta? A new table would have made much more sense (and would have been even more performant)...
